testsuite/tests/typing-misc/build_as_type.ml

```diff
@@ -0,0 +1,112 @@
+(* TEST
+   * expect
+*)
+
+let f = function ([] : int list) as x -> x ;;
+[%%expect{|
+Line 1, characters 8-42:
+1 | let f = function ([] : int list) as x -> x ;;
+            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Warning 8: this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+_::_
```

Comment:
Nitpick: I would silence the warning with a `| _ -> assert false` to have the expect output more readable.

(Same in other cases below.)